lib/ukboot: Introduce early init bootstage

michpappas opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [`arm64`, `x86_64`]
 - Platform(s): [`kvm`, `xen`]
 - Application(s): [N/A]

### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

This series introduces an `early_init` bootstage in `libukboot` executed as soon as the boot protocol passes control to the platform. Libraries can execute in `early_init` by registering functions to `uk_boot_earlytab`, a minimal variant of `uk_inittab`. 

The kernel command line parsing is moved to the earliest priority of `early_init` to provide command line parameters to the rest of the functions registered with earlytab. As all parsing is handled internally by `uk_boot_early_init()`, `ukplat_entry_argp()` is deprecated. Platforms are required to execute `uk_boot_early_init()` before jumping to the entry point of `libukboot`, now renamed to `uk_boot_entry()`.

Bootinfo mrd coalescing is moved to the penultimate priority of the earlytab. Coalescing at the boot protocol is deprecated. The latest priority of earlytab is reserved for post-coalesce operations.

As command line parsing happens before coalescing, it is not possible to copy the command line into a dynamically allocated mrd. `UKPLAT_MEMRT_CMDLINE` and its registration by the boot protocol is deprecated. The boot protocols are updated to provide a direct reference to the command line via `bi->cmdline`. Similarly, copying the command line into a mutable buffer by the platform is deprecated, as that is now internally managed by `uk_boot_early_init()`.

Platforms  (`plat/kvm` & `plat/xen`) are adapted and cleaned up as needed.

GitHub-Depends-On: #1471

michpappas committed:
plat/kvm: Adapt to libukboot early init

Update plat/kvm for newly introduced libukboot early init.

arm64:
- Deprecate kvm/arm64/early_init.c and calls.

arm64 & x86_64:
- Remove ukplat_memregion_list_coalesce() from boot protocols.
  Coalescing is now part of libukboot early init.
- Remove cmdline_init() from kvm/arm64 and kvm/x86_64. Command line
  initialization is now performed of libukboot early init.
- Call uk_boot_early_init() from _ukplat_entry()
- Remove call to ukplat_entry_argp() and jump directly to
  uk_boot_entry()

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/common/bootinfo: Remove mrd allocation of cmdline

Remove cmdline mrd allocation from bootinfo_fdt, as this requirement
has been deprected by libukboot early init.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat: Move SIPI vector mrd allocation to lcpu_mp_init

Move the SIPI vector mdr allocation from the boot protocol
to lcpu_arch_mp_init(). Notice that lcpu_arch_mp_init() must
be called after mrd coalescing.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/kvm/x86: Remove mrd allocation of cmdline

Remove cmdline mrd allocation from the efi, lxboot, multiboot boot
protocols as this requirement has been deprected by libukboot early
init.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/common: Deprecate UKPLAT_MEMRT_CMDLINE

Deprecate UKPLAT_MEMRT_CMDLINE. Passing an cmdline mrd has been
deprecates by libukboot early init.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/xen/arm64: Adapt to libukboot early init

Remove cmdline mrd allocation and call uk_boot_early_init().

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/xen/x86_64: Adapt to libukboot early init

Remove cmdline mrd allocation and call uk_boot_early_init().
Replace call to ukplat_entry_argp() with call to uk_boot_entry().

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/ukboot: Rename ukplat_entry() to uk_boot_entry()

Migrate definition of ukplat_entry() from uk/plat/bootstrap.h
to boot.h and rename to uk_boot_entry().

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/ukboot: Introduce uk_boot_earlytab

uk_boot_earlytab is a simplified version of uk_inittab for libraries
that need to register functions at the early boot stage. Libraries
can register with the earlytab using the uk_boot_earlytab_register()
macro.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/ukboot: Introduce early init bootstage

Introduce uk_boot_early_init(). This implements an early init
bootstage based on uk_boot_earlytab.

Register cmdline parsing at UK_PRIO_EARLIEST to provide cmdline args
to functions registered into the earlytab.

Deprecate ukplat_entry_argp(). Platforms must now call
uk_boot_early_init() to parse the command line arguments.

As mrd allocation must happen after coalesce, but cmdline parsing must
happen at UK_PRIO_EARLY (before coalesce), deprecate the requirement
for boot protocols to pass a cmdline mrd. Also deprecate the
requirement that platforms provide mutable copy of the cmdline mrd,
as copying is internally performed by the cmdline parsing of early
init into newly introduced buffer of configurable size
(CONFIG_LIBUKBOOT_CMDLINE_MAX_LEN).

Register mrd coalescing as the penultimate priority and reserve
UK_PRIO_LATEST for post-coalesce functions.

Register parsing of uklibparam at UK_PRIO_EARLIEST to provide the
rest of earlytab entries with early cmdline. Notice that for the
time being this has the side-effect of information printed by
libuklibparam to be lost, as earlycon is registered to a later
priority. This will be resolved once libukprint is introduced, as
that library will provide a console buffer.

Checkpatch-Ignore: AVOID_EXTERNS
Checkpatch-Ignore: MACRO_ARG_REUSE
Checkpatch-Ignore: STRNCPY
Signed-off-by: Michalis Pappas <michalis@unikraft.io>

mogasergiu commented on lib/ukboot/include/uk/boot/earlytab.h:
```diff
@@ -0,0 +1,48 @@
+#define __UK_EARLYTAB_ENTRY(init_fn, prio)				\
+	static const struct uk_boot_earlytab_entry			\
+	__used __section(".uk_boot_earlytab" #prio) __align(8)		\
+		__uk_boot_earlytab ## prio ## _ ## init_fn = {		\
```
is this overindented?

michpappas replied:
Very much so 👍🏼

mogasergiu commented on plat/common/include/uk/plat/common/common.lds.h:
```diff
@@ -129,6 +129,14 @@
 	}								\
 	uk_inittab_end = .;
 
+#define EARLYTAB_SECTION						\
```
Ditto as in #1471. Additionally, any reason this is not named something like `BOOT_EARLYTAB_SECTION`? As in, the others seem to be defined in such a manner that just the "UK" :british_indian_ocean_territory: is missing.
```C
#define EARLYTAB_SECTION						\
	.uk_boot_earlytab :						\
	{								\
		uk_boot_earlytab_start = .;				\
		KEEP(*(SORT_BY_NAME(.uk_boot_earlytab[0-9])))		\
		uk_boot_earlytab_end = .;				\
	}
```

michpappas replied:
I'm moving this to an library-internal linker script. I'll keep the `uk_` prefix, as it seems to be used elsewhere too:
```
 Section to Segment mapping:
  Segment Sections...
   00     .text 
   01     .rodata .uk_thread_inittab .uk_libinfo .uk_posix_clonetab .eh_frame .eh_frame_hdr .uk_ctortab .uk_inittab .uk_eventtab 
   02     .data .uk_libparam_params .posix_socket_driver_list .uk_bootinfo .bss 
   03     .tdata .tbss 
   04     .tdata 
   05     
```

mogasergiu commented on include/uk/plat/bootstrap.h:
```diff
@@ -50,19 +50,7 @@ extern "C" {
  * @param argc Number of arguments
```
Hmm, maybe also delete the `@param` comments here?

michpappas replied:
Ah yes, but this goes away altogether in a later commit in this PR.

mogasergiu commented on lib/ukboot/early_init.c:
```diff
@@ -0,0 +1,128 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
```
2024

michpappas replied:
Damn

mogasergiu commented on lib/ukboot/early_init.c:
```diff
@@ -0,0 +1,128 @@
+#include <uk/plat/common/bootinfo.h>
+#include <uk/prio.h>
+
+#ifdef CONFIG_LIBUKLIBPARAM
```
`#if`?

mogasergiu commented on lib/ukboot/boot.c:
```diff
@@ -111,6 +107,9 @@
 #include <uk/intctlr.h>
 #endif /* CONFIG_LIBUKINTCTLR */
 
+extern char *argv[CONFIG_LIBUKBOOT_MAXNBARGS];
```
I feel like these names are a bit "established"? Maybe we should just name them as `boot_argv`/`boot_argc` or something? I am fine with this as well but it's just that `argv` and `argc` sound like something other components may want to unexpectedly define for themselves.

michpappas replied:
Yes,  I thought about that too. But then again these vars really hold `argc` and `argv`. Either way, I don't have a strong opinion.

mogasergiu commented on lib/ukboot/early_init.c:
```diff
@@ -0,0 +1,128 @@
+static char cmdline[CONFIG_LIBUKBOOT_CMDLINE_MAX_LEN];
+static __sz cmdline_len;
+
+#ifdef CONFIG_LIBUKLIBPARAM
```
`#if`?

mogasergiu commented on lib/ukboot/early_init.c:
```diff
@@ -0,0 +1,128 @@
+#ifdef CONFIG_LIBUKLIBPARAM
+static int init_libparam(int argc, char *argv[])
+{
+	int rc;
```
Nitpick: a newline afterwards?

Perhaps some assertions for `argv` as well? Especially since `argv[1]` is there.

michpappas replied:
I'll add an assertion for `argv`.

mogasergiu commented on lib/ukboot/early_init.c:
```diff
@@ -0,0 +1,128 @@
+static __sz cmdline_len;
+
+#ifdef CONFIG_LIBUKLIBPARAM
+static int init_libparam(int argc, char *argv[])
```
See, you are using these names as arguments already. I understand the compiler is smart and stuff but I think it could create confusion.

michpappas replied:
It's actually the scope of identifiers as defined by the C standard. Anyway, I can rename them.

mogasergiu commented on lib/ukboot/early_init.c:
```diff
@@ -0,0 +1,128 @@
+		 */
+		/* Run a second pass for parsing */
+		rc = uk_libparam_parse(argc - 1, &argv[1], UK_LIBPARAM_F_USAGE);
+		if (rc < 0) /* go down on errors (including USAGE) */
```
`unlikely`?

michpappas replied:
I just migrated that from `boot.c` verbatim. Will add.

mogasergiu commented on lib/ukboot/early_init.c:
```diff
@@ -0,0 +1,128 @@
+	 * NOTE: argv[0] contains the kernel/program name that we need to
+	 *       hide from the parser.
+	 */
+	rc = uk_libparam_parse(argc - 1, &argv[1], UK_LIBPARAM_F_SCAN);
```
Nitpick: perhaps `rc` is not really the proper name for the variable. Maybe `arg_idx`?

michpappas replied:
Again, that's migrated from `boot.c`.

mogasergiu commented on lib/ukboot/early_init.c:
```diff
@@ -0,0 +1,128 @@
+static int uk_boot_early_init_parse_cmdl(struct ukplat_bootinfo *bi)
+{
+	const char *cmdl;
+	int rc __maybe_unused;
```
AM I SEEING THIS RIGHT? IS THIS.... NOT REVERSE XMAS TREE? :monocle_face: :laughing:

## UNACCEPTABLE

michpappas replied:
It's okay. We don't have an established rule for that anyway.

mogasergiu commented on lib/ukboot/early_init.c:
```diff
@@ -0,0 +1,128 @@
+
+	UK_ASSERT(bi);
+
+	bi = ukplat_bootinfo_get();
```
Why do we need to fetch it again? Misplaced getter maybe?

michpappas replied:
Yes, it's been forgotten. I'll remove 👍🏼

mogasergiu commented on lib/ukboot/early_init.c:
```diff
@@ -0,0 +1,128 @@
+		cmdline_len = bi->cmdline_len;
+	}
+
+	if (cmdline_len > ARRAY_SIZE(cmdline))
```
`unlikely`? And why `ARRAY_SIZE` instead of `sizeof`?

michpappas replied:
`unlikely` ok. Why *not* `ARRAY_SIZE`?

mogasergiu replied:
I like `sizeof` better because it is built-in and makes it clear we are dealing with a string. Maybe it is just personal preference though?

michpappas replied:
Yeah, I think so as others may argue that it's safer to always use `ARRAY_SIZE` regardless of the underlying type.

mogasergiu replied:
Ok fine. I just find it unusual being used for strings.

mogasergiu commented on lib/ukboot/early_init.c:
```diff
@@ -0,0 +1,128 @@
+	if (cmdline_len > ARRAY_SIZE(cmdline))
+		return -EINVAL;
+
+	strncpy(cmdline, cmdl, ARRAY_SIZE(cmdline));
```
Ditto

mogasergiu commented on lib/ukboot/early_init.c:
```diff
@@ -0,0 +1,128 @@
+	argc = uk_argnparse(cmdline, cmdline_len, &argv[0],
+			    CONFIG_LIBUKBOOT_MAXNBARGS);
+
+#ifdef CONFIG_LIBUKLIBPARAM
```
`#if`?

mogasergiu commented on lib/ukboot/early_init.c:
```diff
@@ -0,0 +1,128 @@
+}
+
+/* Parse cmdline parameters at the earlies priority
+ * to make them available the rest of early init
```
```diff
- Parse cmdline parameters at the earlies priority
+ Parse cmdline parameters at the earliest priority
- to make them available the rest of early init
+ to make them available to the rest of early init
```

mogasergiu commented on lib/ukboot/include/uk/boot.h:
```diff
@@ -9,12 +9,18 @@
+/**
+ * Executes the early_init boot stage
+ */
+void uk_boot_early_init(struct ukplat_bootinfo *bi);
```
Maybe a `@param` arg comment as well?

mogasergiu commented on plat/common/bootinfo_fdt.c:
```diff
@@ -85,38 +85,23 @@ static void fdt_bootinfo_mem_mrd(struct ukplat_bootinfo *bi, void *fdtp)
 }
 
-static void fdt_bootinfo_cmdl_mrd(struct ukplat_bootinfo *bi)
+static void fdt_bootinfo_cmdl_init(struct ukplat_bootinfo *bi, void *fdtp)
```
In the commit msg: `s/rmd/mrd/`.
Also maybe `const void *fdtp`?

michpappas replied:
> In the commit msg: s/rmd/mrd/.

Ok

>  Also maybe const void *fdtp?

I'm following the original author's convention 😉  Saying that, `fdtp` is redundant in most of these functions, as it can be obtained from the `bi` parameter already, so I can add a commit to clean that up.

mogasergiu replied:
Makes sense.

mogasergiu reviewed (changes requested):
Greetings, mister sir! :wave: :monocle_face: 
Very happy about this early tab addition. One step closer to welcoming ACPI into a beautiful home! Also cool how you took care of Xen as well :pray:

mogasergiu commented on lib/ukboot/earlytab.lds.S:
```diff
@@ -0,0 +1,9 @@
+SECTIONS {
```
License?

mogasergiu commented on lib/ukboot/include/uk/boot/earlytab.h:
```diff
@@ -0,0 +1,48 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
```
2024

michpappas replied:
Some people never learn...

mogasergiu commented on lib/ukboot/early_init.c:
```diff
@@ -0,0 +1,137 @@
+	}
+
+	if (unlikely(cmdline_len > ARRAY_SIZE(cmdline)))
+		return -EINVAL;
```
Maybe an error message. Why `EINVAL`? Would `E2BIG` fit better here?

michpappas replied:
Nice, I forgot about `E2BIG`.

mogasergiu commented on lib/ukboot/early_init.c:
```diff
@@ -0,0 +1,137 @@
+				 CONFIG_LIBUKBOOT_MAXNBARGS);
+
+#if CONFIG_LIBUKLIBPARAM
+	rc = init_libparam(boot_argc, boot_argv);
```
Why not just `return init_libparam`. Either that or I would go as you have now and add an additional error message. The more the better (I know strings increase the size of the binary, but I think it is not something that big...)

mogasergiu commented on lib/ukboot/early_init.c:
```diff
@@ -0,0 +1,137 @@
+uk_boot_earlytab_register(uk_boot_early_init_parse_cmdl, UK_PRIO_EARLIEST);
+
+/* Reserve UK_PRIO_LATEST for post-coalesce operations */
+uk_boot_earlytab_register(uk_boot_early_init_mrd_coalesce,
```
I wonder if there is an easy way to ensure the fact that no one allocates mrd's before coalesce.

michpappas replied:
I guess we could add some logic for that, as well as to enforce a single call of coalesce. If bootinfo becomes part of `libukboot` that makes sense, otherwise if it becomes a standalone lib it seems to add too much policy so I would probably make that behavior configurable. We can keep it in mind as a feature.

mogasergiu commented on plat/common/bootinfo_fdt.c:
```diff
@@ -199,21 +184,7 @@ void ukplat_bootinfo_fdt_setup(void *fdtp)
-		ukplat_bootinfo_crash("Invalid bootinfo");
-
-	fdt_bootinfo_cmdl_mrd(bi);
+	bi->dtb = (__u64)fdtp;
```
>> I'm following the original author's convention 😉 Saying that, fdtp is redundant in most of these functions, as it can be obtained from the bi parameter already, so I can add a commit to clean that up.

> Makes sense.

Oh but wait it seems to be assigned later

michpappas replied:
Yes, ut that doesn't stop us from assigning it as soon as we the sanity check succeeds.
```diff
diff --git a/plat/common/bootinfo_fdt.c b/plat/common/bootinfo_fdt.c
index ddf88e4e9..e9953438d 100644
--- a/plat/common/bootinfo_fdt.c
+++ b/plat/common/bootinfo_fdt.c
@@ -180,11 +180,11 @@ void ukplat_bootinfo_fdt_setup(void *fdtp)
        if (unlikely(fdt_check_header(fdtp)))
                ukplat_bootinfo_crash("Invalid DTB");
 
+       bi->dtb = (__u64)fdtp;
+
        fdt_bootinfo_fdt_mrd(bi, fdtp);
        fdt_bootinfo_mem_mrd(bi, fdtp);
        fdt_bootinfo_initrd_mrd(bi, fdtp);
 
        fdt_bootinfo_cmdl_init(bi, fdtp);
-
-       bi->dtb = (__u64)fdtp;
 }
```

What do you think?

mogasergiu replied:
Ok makes sense. It still feels awkward but I dislike having unnecessary secondary arg. :cold_sweat:

michpappas commented:
Rebased to the latest state of #1471

mogasergiu commented on lib/ukboot/early_init.c:
```diff
@@ -0,0 +1,143 @@
+	 * NOTE: argv[0] contains the kernel/program name that we need to
+	 *       hide from the parser.
+	 */
+	rc = uk_libparam_parse(*argc - 1, &argv[1], UK_LIBPARAM_F_SCAN);
```
Maybe we should assert `argc` as well?

mogasergiu commented on plat/kvm/x86/multiboot.c:
```diff
@@ -71,18 +71,6 @@ void multiboot_entry(struct lcpu *lcpu, struct multiboot_info *mi)
 	if (mi->flags & MULTIBOOT_INFO_CMDLINE) {
 		if (mi->cmdline) {
 			cmdline_len = strlen((const char *)(__uptr)mi->cmdline);
```
Maybe we can get rid of `cmdline_len` variable entirely and directly assign `bi->cmdline_len` to this `strlen`.

michpappas replied:
Actually IMO these are different things: `bi->cmdline_len` reflects what was provided by the boot protocol and it also refers to the length of `bi->cmdline`. If we update `bi->cmdline_len` it would make sense to also update `bi->cmdline`, but I'm not sure there is value on that.

michpappas replied:
Wait sorry somehow I thought you referred to `uk_boot_early_init_parse_cmdline()`. The proposed change makes sense in multiboot, I'll push an update.

mogasergiu commented on plat/kvm/x86/setup.c:
```diff
@@ -24,57 +25,12 @@
+	int rc;
 
-	_libkvmplat_init_console();
+	uk_boot_early_init(bi);
```
Since this has assertions, maybe we should move it below `traps_table_init()`?

mogasergiu commented on plat/kvm/x86/setup.c:
```diff
@@ -134,5 +85,13 @@ void _ukplat_entry(struct lcpu *lcpu, struct ukplat_bootinfo *bi)
+	 * previous instruction pointer as undefined, so that debuggers or
+	 * profilers stop unwinding here.
+	 */
+	ukarch_cfi_unwind_end();
```
Are you sure this is ok to use before the `jmp`? I would still play it safe and use an intermediate function.

michpappas replied:
Sure, I'm not knowledgeable enough on x86's cfi so I'll revert just in case.

mogasergiu commented on plat/kvm/x86/setup.c:
```diff
@@ -134,5 +85,13 @@ void _ukplat_entry(struct lcpu *lcpu, struct ukplat_bootinfo *bi)
+	 */
+	ukarch_cfi_unwind_end();
+	lcpu_arch_jump_to(bstack, uk_boot_entry);
+	ukplat_lcpu_halt();
```
I know this is from before. Maybe we should have `UK_BUG()` instead?

michpappas replied:
For sure. I didn't like that `ukplat_lcpu_halt()` either.

mogasergiu commented on plat/xen/arm/setup64.c:
```diff
@@ -185,16 +185,7 @@ static inline void _get_cmdline(struct ukplat_bootinfo *bi)
-
-	bi->cmdline = (__u64)cmdline;
+	bi->cmdline = (__u64)cmdl;
 	bi->cmdline_len = cmdline_len;
```
Ditto about `cmdline_len` here. Maybe we can use `len` directly.

mogasergiu commented on plat/xen/x86/setup.c:
```diff
@@ -183,49 +184,10 @@ static int _init_mem(struct ukplat_bootinfo *const bi)
 {
 	const char bl[] = "Xen";
 	const char bp[] = "PVH";
-	struct ukplat_bootinfo *bi;
```
Should we assert for `bi`?

mogasergiu commented on plat/xen/arm/setup64.c:
```diff
@@ -169,7 +170,6 @@ static inline void _get_cmdline(struct ukplat_bootinfo *bi)
 {
 	int fdtchosen, len;
 	const char *fdtcmdline, *cmdl;
-	char *cmdline;
```
Ditto for the `bi` here.

mogasergiu reviewed (approved):
Appreciate these changes :pray:. They will serve the reincarnation of our lord and savior ACPI well in the future :godmode:.

Reviewed-by: Sergiu Moga <sergiu@unikraft.io>

michpappas commented:
Rebased from #1471, no functional changes.

skuenzer commented on lib/ukboot/include/uk/boot/earlytab.h:
```diff
@@ -0,0 +1,48 @@
+
+typedef int (*uk_boot_earlytab_func_t)(struct ukplat_bootinfo *bi);
+
+struct uk_boot_earlytab_entry {
```
Would we need the struct for future extension? Do you have something already in mind? Otherwise, I think we can handle this table similar as standard constructors.

michpappas replied:
Since we align the struct to 8 bytes I don't think it makes any difference. Below you can see two entries registered from two different libraries:
```
Hex dump of section '.uk_boot_earlytab':
  0x001621b8 10861200 00000000 c0851200 00000000 ................
``` 
It's more of a matter of habit / good practice in case we need to extend later. Saying that, I don't have an extension in mind atm, so if you have a strong preference we can easily switch to functions.

skuenzer replied:
This is actually true, since it is a single-field `struct` with a pointer, you should be able to cast it to a function pointer even: no real difference in memory. I am also undecided which version is better, normally I tend to minimalism but here we have zero runtime overhead due to the `struct`. Could be worth keeping it. 🤔 Let's keep it, as is.

skuenzer commented on lib/ukboot/boot.c:
```diff
@@ -238,8 +238,7 @@ static struct uk_alloc *heap_init()
 
-/* defined in <uk/plat.h> */
-void ukplat_entry(void)
+void uk_boot_entry(void)
```
There is a `_` too much in the commit title in `uk_plat_entry`, should be `ukplat_entry`.

michpappas replied:
Actually `_ukplat_entry` is correct, I am referring to the platform entry, not the boot entry: https://github.com/unikraft/unikraft/blob/staging/plat/kvm/x86/setup.c#L72 🙂

skuenzer replied:
What I meant is the title of this commit that does this change:
```
lib/ukboot: Rename uk_plat_entry() to uk_boot_entry() 
```
Shouldn't it be?:
```
lib/ukboot: Rename ukplat_entry() to uk_boot_entry() 
```

michpappas replied:
That one yes 👍🏼

michpappas commented:
Rebased from `staging` to pick up #1471

skuenzer commented on lib/ukboot/boot.c:
```diff
@@ -443,7 +396,8 @@ void ukplat_entry(int argc, char *argv[])
 #else /* CONFIG_LIBUKBOOT_MAINTHREAD */
 	m = uk_sched_thread_create_fn2(s, main_thread,
-				       (void *)((long)argc), (void *)argv,
+				       (void *)((long)boot_argc),
```
Hum, shouldn't it be `ictx.cmdline.argc` and `ictx.cmdline.argv` here?

skuenzer replied:
The idea was that init functions would be able to replace/modify argc and argv.

skuenzer commented on lib/ukboot/early_init.c:
```diff
@@ -0,0 +1,150 @@
+/* Mutable version of the cmdline. This is tokenized by uk_argnparse().
+ * Keep this as a global as `argv` points to tokens in this string.
+ */
+static char cmdline_argv[CONFIG_LIBUKBOOT_CMDLINE_MAX_LEN];
```
Could we find a different name than `cmdline_argv`. With `_argv` I am biased to read `char *argv[]` 🤔

skuenzer replied:
Plus, see my questions below.

michpappas replied:
Will rename to `cmdline_argparse`

skuenzer commented on lib/ukboot/early_init.c:
```diff
@@ -0,0 +1,150 @@
+
+	/* Break down cmdline to argv */
+	strncpy(cmdline_argv, cmdl, ARRAY_SIZE(cmdline));
+	cmdline_argv[cmdline_len] = 0;
```
I'd prefer to stay with the target data type: Can we place `'\0'` here, instead of `0`?

skuenzer commented on lib/ukboot/early_init.c:
```diff
@@ -0,0 +1,150 @@
+	bi->cmdline = (__u64)cmdline;
+
+	/* Break down cmdline to argv */
+	strncpy(cmdline_argv, cmdl, ARRAY_SIZE(cmdline));
```
Shouldn't the size be derived from `cmdline_argv`?: `ARRAY_SIZE(cmdline_argv)`

skuenzer commented on lib/ukboot/early_init.c:
```diff
@@ -0,0 +1,150 @@
+	}
+
+	/* Update bi->cmdline */
+	strncpy(cmdline, cmdl, ARRAY_SIZE(cmdline));
```
Why do we do two copies of `cmdl`? Wouldn't `boot_argv` and `boot_argc` be good enough to keep only?

michpappas replied:
`cmdline` is required for preserving `bi->cmdline` as the boot protocol originally points `bi->cmdline` to memory that, depending on the boot protocol, it may get unmapped later (and we print it with `ukplat_bootinfo_print()`). This is not new, we were doing it already in platform code for all platforms.

`cmdline_argv` is used for parsing, as it is updated in-place with `\0` by `uk_argnparse()`. This is not new either, it is just that now the two copies stand out as they're centralized in `early_init()`.

See also: https://github.com/unikraft/unikraft/pull/1472/commits/57aaa2df8d9335b04f17839968b7aefdcefce9b8#diff-02014d06ce1464d989f9bc78a64dc2159ff644d8cc71f1882a75730c55076895R21

skuenzer replied:
Okay, the question is then rather if there is a use case for keeping `bi->cmdline`? However, I think we can resolve that in a different PR.

To explain a bit better my motivation: if we would provide an ukargparse variant that would work cpy and parse a buffer, we could use that function directly to read from the original cmdline and copy the parsed arguments directly over.

michpappas replied:
@skuenzer, you right.  What I propose is:

1. Formalize the call to `ukplat_bootinfo_print()` as Unikraft's "hello" by moving it to early_init before we iterate the earlytab. With the addition of the logbuf it will be printed once the console is up.
2. Set `bi->cmdline` to NULL as soon as we have parsed the cmdline. That will prevent manual calls to `ukplat_bootinfo_print()` - often used for debugging mrds - from causing an abort.
3. Remove the immutable copy of the cmdline.

I pushed an update with these changes.

skuenzer commented on lib/ukboot/include/uk/boot.h:
```diff
@@ -19,14 +19,25 @@ extern "C" {
+ * bootstrapping. Before calling this function a platform must
+ * have called uk_boot_early_init().
+ *
+ * Notice: Once all platform code registers to earlyrtab this
```
- Typo: `earlyrtab`
- I would change `Notice:` to `TODO:` or `NOTE:`

skuenzer commented on plat/kvm/arm/setup.c:
```diff
@@ -231,5 +194,6 @@ void __no_pauth _ukplat_entry(void)
 	uk_pr_info("Switch from bootstrap stack to stack @%p\n", bstack);
 
-	lcpu_arch_jump_to(bstack, _ukplat_entry2);
+	lcpu_arch_jump_to(bstack, uk_boot_entry);
```
I guess all calls to `uk_boot_entry` imply adding `lib/ukboot` as dependency (`select`) to the corresponding platform's `Config.uk`.

skuenzer commented on plat/xen/arm/setup64.c:
```diff
@@ -31,6 +31,7 @@
 #include <common/events.h>
 #include <common/console.h>
 #include <xen/xen.h>
+#include <uk/boot.h>
```
Same here: A `lib/ukboot` dependency

skuenzer commented on plat/xen/x86/setup.c:
```diff
@@ -71,6 +71,7 @@
 #include <uk/config.h>
 #include <uk/print.h>
 #include <uk/assert.h>
+#include <uk/boot.h>
```
Another `lib/ukboot` dependeny for `Config.uk`

oleksiimoisieiev commented on plat/xen/arm/setup64.c:
```diff
@@ -167,35 +168,23 @@ static void *map_fdt(paddr_t device_tree)
 		if (fdtcmdline) {
-			cmdl = fdtcmdline;
-			cmdline_len = len;
+			bi->cmdline = fdtcmdline;
```
I think this a bit unsafe. fdt_getprop call is used to get fdtcmdline property. Which returns address of the fdtproperty->data member. fdt_getprop is the part of fdt_ro, which means that this data is ro. So if we put this pointer to __u64 cmdline in the bi structure - there is a potential risk that this memory may be changed which may lead to unexpected consequences, such as wrong cmdline in the device tree blob or even fdt corruption.
That's why cmdline was copied on the previous version.
I think in this case either copy of cmdline should be done or __u64 cmdline in bi structuure should be const. What do you think?

michpappas replied:
You're right, and as a matter of fact we do copy the cmdline, but now we have formalized that behavior for all platforms (as every platform was copying its cmdline anyway) at the implementation of `uk_boot_early_init_parse_cmdl()`: https://github.com/unikraft/unikraft/pull/1472/commits/9930d23655e811a430ba49b8d3f67a58df8eba19#diff-02014d06ce1464d989f9bc78a64dc2159ff644d8cc71f1882a75730c55076895R98.
So essentially nothing has changed.

oleksiimoisieiev commented on plat/xen/x86/setup.c:
```diff
@@ -183,63 +184,32 @@ static int _init_mem(struct ukplat_bootinfo *const bi)
 		UK_CRASH("Failed to initialize memory\n");
 
-	_libxenplat_x86bootinfo_setup_cmdl(bi);
+	bi->cmdline_len = strlen((char *)HYPERVISOR_start_info->cmd_line);
```
I would use strnlen((char *)HYPERVISOR_start_info->cmd_line, CONFIG_LIBUKBOOT_CMDLINE_MAX_LEN);

michpappas replied:
That's a good point, however we don't export `CONFIG_LIBUKBOOT_CMDLINE_MAX_LEN` outside `libukboot`. What we do instead is expect that the boot protocol sets `bi->cmdline` and `bi->cmdline_len` to memory it controls, and manage the copy - including the check to the size of the buffers controlled by libukboot - internally in `libukboot`. That provides some separation between the responsibilities of each component involved with boot, and avoids having the boot protocol make any assumptions about `libukboot` internals.

oleksiimoisieiev commented on plat/xen/x86/setup.c:
```diff
@@ -183,63 +184,32 @@ static int _init_mem(struct ukplat_bootinfo *const bi)
 {
+	struct ukplat_bootinfo *bi;
+
+	bi = ukplat_bootinfo_get();
```
`struct ukplat_bootinfo *bi = ukplat_bootinfo_get();` ??

michpappas replied:
That's because of the coding convention discouraging this style: https://unikraft.org/docs/contributing/coding-conventions#variable-definitions

oleksiimoisieiev commented on lib/ukboot/boot.c:
```diff
@@ -403,8 +356,8 @@ void ukplat_entry(int argc, char *argv[])
 
-	ictx.cmdline.argc = argc;
-	ictx.cmdline.argv = argv;
+	ictx.cmdline.argc = boot_argc;
```
Probably some check needs to be here to ensure that boot_argv and boot_argc are already processed at this stage.

michpappas replied:
Good point, I'll add an assertion 👍🏼

oleksiimoisieiev commented on lib/ukboot/early_init.c:
```diff
@@ -0,0 +1,150 @@
+
+	/* Break down cmdline to argv */
+	strncpy(cmdline_argparse, cmdl, ARRAY_SIZE(cmdline_argparse));
+	cmdline_argparse[cmdline_len] = '\0';
```
As I can see from the code - cmdline_argparse is not used. Here are search results:
```
lib/ukboot/early_init.c:static char cmdline_argparse[CONFIG_LIBUKBOOT_CMDLINE_MAX_LEN];
lib/ukboot/early_init.c:        strncpy(cmdline_argparse, cmdl, ARRAY_SIZE(cmdline_argparse));
lib/ukboot/early_init.c:        cmdline_argparse[cmdline_len] = '\0';
````

michpappas replied:
Thanks for catching that, it should be passed to `uk_argnparse()`. This is a regression introduced in the last update of the PR. Will fix 👍🏼

andreistan26 reviewed (approved):
After all the heavy lifting was done, nothing new from my side, additions look good!

Reviewed-by: Andrei Stan <andreistan2003@gmail.com>

michpappas commented:
@oleksiimoisieiev I addressed your comments (see replies inline as well as https://github.com/unikraft/unikraft/compare/8babcb3e87b5df4ff540f0bdb020ea0ab70fb2e3..b11bdd5176da6dc2d58418ac30360a1fd5595208) and additionally fixed a bug during parsing that caused incorrect arguments passed to the application.

oleksiimoisieiev commented:
LGTM 

Reviewed-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>

skuenzer reviewed (approved):
Great work! Thanks a lot!

Reviewed-by: Simon Kuenzer <simon@unikraft.io>
Approved-by: Simon Kuenzer <simon@unikraft.io>
